std/stdio.d

```diff
@@ -5457,18 +5455,53 @@ private struct ReadlnAppender
     }
 }
 
-// Private implementation of readln
-private size_t readlnImpl(FILE* fps, ref char[] buf, dchar terminator, File.Orientation orientation)
+private struct LockedFile
 {
-    version (DIGITAL_MARS_STDIO)
+    private @system FILE* fps;
+
+    this(FILE* fps) @trusted
     {
+        this.fps = fps;
         _FLOCK(fps);
-        scope(exit) _FUNLOCK(fps);
+    }
+
+    @disable this();
+    @disable this(this);
+    @disable void opAssign(LockedFile);
 
-        /* Since fps is now locked, we can create an "unshared" version
-         * of fp.
-         */
-        auto fp = cast(_iobuf*) fps;
+    // these use unlocked fgetc calls
+    // Since fps is now locked, we can cast away shared
+    @trusted fgetc() { return _FGETC(cast(_iobuf*) fps); }
+    @trusted fgetwc() { return _FGETWC(cast(_iobuf*) fps); }
+
+    ~this() @trusted
+    {
+        _FUNLOCK(fps);
+    }
+}
+
+@safe unittest
+{
+    void f() @safe
+    {
+        FILE* fps;
+        auto lf = LockedFile(fps);
+        static assert(!__traits(compiles, lf = LockedFile(fps)));
+        version (ShouldFail)
+        {
+            lf.fps = null; // error with -preview=systemVariables
+        }
+    }
+}
+
+// Private implementation of readln
+private size_t readlnImpl(FILE* fps, ref char[] buf, dchar terminator, File.Orientation orientation) @safe
+{
+    version (DIGITAL_MARS_STDIO)
+    {
+        auto lf = LockedFile(fps);
+        // Since fps is now locked, we can cast away shared
+        auto fp = (() @trusted => cast(_iobuf*) fps)();
```

Comment:
this shouldn't be done at all, use the `LockedFile`. Just add any needed trusted methods to the struct.

If all trusted operations are encapsulated in the struct, then you don't have to worry about this function in terms of safety (you do right now, because `@system` data isn't yet a thing, but eventually).

Reply:
> Just add any needed trusted methods to the struct

Well, they're fields:
```
stdio.d:5509: if (__fhnd_info[fp._file] & FHND_WCHAR)
stdio.d:5541: else if (fp._flag & _IONBF)
stdio.d:5565: int u = fp._cnt;
stdio.d:5566: char* p = fp._ptr;
stdio.d:5568: if (fp._flag & _IOTRAN)
stdio.d:5610: fp._cnt -= i;
stdio.d:5611: () @trusted { fp._ptr += i; }();
```
Maybe I could add an `opDispatch`.

> If all trusted operations are encapsulated in the struct, then you don't have to worry about this function in terms of safety

I already need to call a trusted lambda to call `free`. I don't really see what's wrong with immediately calling a trusted lambda, the `@trusted`ness is not being hidden. It's interface is safe at that point in the execution.

Reply:
OK, I suppose if there's a `lf.destroy` call added then the `fp` is not safe.

Reply:
Most of this does not need to happen on the structure itself, use locals (it's actually faster anyway to manipulate local pointers/counters than use an indirection, from my experience in iopipe). Also, if needed you can encapsulate the whole functionality into the struct itself, or make this section a trusted lambda.

Yeah, I'm looking at the code now, and actually I think this isn't getting the picture across, I'm sorry, trusted code is just so hard to get right. The opDispatch is incorrectly marked safe, because it's treating the pointer as a safe pointer, when it should be a system pointer.

When I get a few minutes, I'll rewrite what I think it should be.

Reply:
@schveiguy I think I've got it right now, thanks for your help. I put all the digital mars stdio code in a single trusted lambda and did the same for the getdelim code too.

Reply:
Looks a lot better! One thing I would recommend, store the actual pointer under a more obscure name, and access via a `@system` accessor, which will enable proper safety checking even without the `@system` data DIP implemented. Currently the simple access of the private member is still considered `@safe` by the compiler, and this will at least involve the enforcement, even if the requirement is artificial.

e.g.:
```d
private struct LockedFile {
   @system private _iobuf* _fp; // obscure name to avoid accidental usage
   @system _iobuf* fp() { return _fp; } // use this in all @trusted calls
   ...
}
```

When the system variable DIP is default, this can be changed to just a field.